Cross-platform patcher

LennardF1989 opened the pull request:
I've updated the patcher to .NET 8, as-is. I touched as little code as possible, I could even (barely) hold myself back from applying proper C# name conventions *shivers*.

I then split the logic into 3 components: UI, CLI and Core. UI is pretty much Windows-only (but could run under Proton), CLI is cross-platform and spits out a native executable for either Windows or Linux.

Workflow has been adjusted to spit out 6 files:
- CLI: Windows/Linux Portable/Standalone
- UI: Windows Portable/Standalone

CLI is pretty small when completely standalone (12MB-ish). 
UI is humongous when standalone (125MB, 50MB compressed). Not sure how I feel about that, but WinForms can't be stripped (yet). UI could potentially be replaced with a different solution.

This will bring the patcher into "modern era", as well as add support for Steam Deck (**test pending**).

CLI also has additional logic applied so it will respect settings from the config, as well as waiting for a HITMAN process before applying and exiting.

LennardF1989 committed:
Updated patcher to .NET 8.0

LennardF1989 committed:
Split the codebase into two executables: UI and CLI

LennardF1989 committed:
Updated patcher workflow

LennardF1989 committed:
Added support for unix file paths
Updated workflow to also build smaller portable executables

LennardF1989 committed:
Updated to actions/setup-dotnet@v4
Fixed wrong filenames for Linux artifacts

RDIL commented:
This is incredible! 🤩 Never thought I would see the day!

RDIL commented:
> UI is humongous when standalone (125MB, 50MB compressed). Not sure how I feel about that, but WinForms can't be stripped (yet). UI could potentially be replaced with a different solution.

Probably controversial idea here - we might not even need a UI. If we have a script that starts the CLI in command prompt and then once the game is patched, it can just say something like "Done! You may now close this."

alex73630 commented:
Thanks Lennard for this amazing work!
If you need help testing Linux/Steam Deck support, I'll gladly help you out

alex73630 commented:
> Probably controversial idea here - we might not even need a UI. If we have a script that starts the CLI in command prompt and then once the game is patched, it can just say something like "Done! You may now close this."

I don't think it's a good idea to completely remove the UI option, knowing some of our users, this will create more problems than shipping a 50MB binary

LennardF1989 commented:
I gave this same more thought, and I might be able to turn the Core-lib into a .netstandard1 package, which would technically allow the UI to be run as good-ol' .net 4.6, but the CLI as .NET 8, while still sharing the same logic AND having the space benefit.

LennardF1989 commented:
Above works. However, using a native Unix process to perform patching causes all kinds of issues, since Linux doesn't have the functions Windows has to apply the patches.

I'll rethink this but I see a few options: get the CLI in top shape for proton, fire and forget.

Or, use a hook dll to perform the patches natively. Benefit of this is also that you could use something like ImGui as a replacement for the UI. To be continued.

LennardF1989 commented:
BOOM! Also appears to work without sudo, which is great.

<img width="1157" alt="image" src="https://github.com/thepeacockproject/Peacock/assets/869973/aef7e333-5d76-4ff7-b86a-e680fd0c9dc0">

LennardF1989 committed:
Added support for both .NET 8 and .NET 4.6.1

LennardF1989 committed:
Added Pinvoke abstractions for Linux

LennardF1989 committed:
Hotfixes after testrun on Steam Deck

RDIL commented on patcher/HitmanPatcher.UI/MainForm.Designer.cs:
```diff
@@ -1,4 +1,4 @@
-﻿namespace HitmanPatcher
+namespace HitmanPatcher
```
[BOM](https://en.wikipedia.org/wiki/Byte_order_mark)/[zwsp](https://en.wikipedia.org/wiki/Zero-width_space)/[nbsp](https://en.wikipedia.org/wiki/Non-breaking_space) added/removed here?

LennardF1989 replied:
Nope, check with Hex editor. I guess it's Githubs way of showing a rename.

LennardF1989 committed:
Added cross-compilation for .NET Framework 4.6 and .NET 8
Replaced embedded resources with custom implementation for proper cross-compilation support

LennardF1989 committed:
Attempt at fixing CodeQL

LennardF1989 committed:
Fixed copy/paste error in codeql.yml

LennardF1989 commented:
Build now outputs a Windows-only .NET Framework 4.6 version, exactly like the original, just through a different build system. 

The .NET 8 version is cross-compiled for Windows and Linux and comes in self-contained and portable versions.

Should be ready for testing.

LennardF1989 commented:
Setup files for Steam Deck: [PeacockSteamDeck.zip](https://github.com/thepeacockproject/Peacock/files/14323372/PeacockSteamDeck.zip)

# Installation
1. Switch to Desktop Mode
2. Open a terminal
3. Create a directory anywhere using `mkdir Peacock` and `cd Peacock` into it
4. Execute the following command (use the "Copy to clipboard"-button):
```bash
wget https://github.com/thepeacockproject/Peacock/files/14323372/PeacockSteamDeck.zip && unzip PeacockSteamDeck.zip && chmod +x ./start.sh && rm PeacockSteamDeck.zip
```
5. Run `./start.sh` once to make sure everything is setup properly.
6. Open a browser and go to http://localhost:3000 to confirm Peacock is running.
7. Now open Steam, add a non-Steam game "Peacock" and choose the `start.sh`. Optionally assign the icon to be `icon.png` inside the Peacock-directory.

# Running
1. Switch back to Gaming Mode.
2. Run the non-Steam game "Peacock". Leave it running like this, do not press B!
3. Instead press the Steam-button on your Steam Deck to bring up your library. Either run Heroic and launch Hitman from there, or run the Steam version.

The game will be patched accordingly and connect to your own Peacock server.

# Closing
Just close the game, switch back to Peacock and exit the application. This will take down both the server and the patcher.*

* This is non-standard script behavior, but I guess Steam keeps track of the processes any process starts and kills all of them whenever you kill the parent process. Which is really nice for this situation!

crazytelli commented:
@LennardF1989 I was able to run the server and confirm in my browser that localhost:3000 is indeed running peacock.

I had to make a slight change in `start.sh` as follows:

`FOLDER_NAME="Peacock-${LATEST_RELEASE}-linux"` as the download file for linux is no longer named `lite`.

The one problem I'm running now is the Patcher is failing when launching hitman.

Here are some screenshots:

![peacock-server-running](https://github.com/thepeacockproject/Peacock/assets/56113536/26a572d2-2417-419e-bab3-22893ad3d50c)


![patcher-failed](https://github.com/thepeacockproject/Peacock/assets/56113536/af6adff4-9331-4ca3-aa12-cbfd6b40fd18)

RDIL commented:
@crazytelli thanks for the heads up, I changed it from lite -> linux in the linux setup repo. Will check on the error later.

LennardF1989 commented:
Which version are you running @crazytelli? Steam, Epic? And if so, which version is reported in the main menu?

crazytelli commented:
> Which version are you running @crazytelli? Steam, Epic? And if so, which version is reported in the main menu?

I`m running the Steam version of Hitman III.

The game version (if I'm understanding what you're asking correctly) - in game -> Options:
- game version: 3.180.0
- server version: 8.15.0

crazytelli commented:
Consistently, never actually managed to launch peacock successfully with this method.

-------- Mensagem original --------
Em 06/04/2024 08:44, Lennard Fonteijn  escreveu:

> And is it consistently failing to patch, or occasionally?
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/thepeacockproject/Peacock/pull/388#issuecomment-2041059573), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ANMDTAEVTNQ5DC2IJBIEZILY37NZLAVCNFSM6AAAAABDHN2WMOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBRGA2TSNJXGM).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

LennardF1989 committed:
Added option for CLI to keep scanning for processes
Fixed issue where legacy GUI also has a console

LennardF1989 commented:
> Consistently, never actually managed to launch peacock successfully with this method.
> […](#)
> -------- Mensagem original -------- Em 06/04/2024 08:44, Lennard Fonteijn escreveu:
> And is it consistently failing to patch, or occasionally? — Reply to this email directly, [view it on GitHub]([#388 (comment)](https://github.com/thepeacockproject/Peacock/pull/388#issuecomment-2041059573)), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ANMDTAEVTNQ5DC2IJBIEZILY37NZLAVCNFSM6AAAAABDHN2WMOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBRGA2TSNJXGM). You are receiving this because you were mentioned.Message ID: ***@***.***>

I've tested this on both the Steam and Epic version, with both the CLI and GUI version, with both the .NET 8 version and legacy version and I cannot reproduce this issue.

I will deploy the latest version to my Steam Deck to confirm everything still works there, too.

LennardF1989 committed:
Fixed patching not working for the Steam version on Linux

LennardF1989 commented:
Found and fixed the issue on the Steam Deck. Was an absolute pain to get my local development environment running there, but I'll make a write up about it soon.

Instructions of https://github.com/thepeacockproject/Peacock/pull/388#issuecomment-1951378211 still apply, however: I made a new version of the zip:
- [PeacockSteamDeck.zip](https://github.com/user-attachments/files/16045984/PeacockSteamDeck.zip)

This one supports updating either Peacock or Node when it's outdated. Peacock will only ever overwrite files, not delete them, so your profile data should be safe. Node is installed based on the .nvm file that comes with Peacock.

New command to install:
```
wget https://github.com/user-attachments/files/16045984/PeacockSteamDeck.zip && unzip PeacockSteamDeck.zip && chmod +x ./start.sh && rm PeacockSteamDeck.zip
```
@crazytelli Can you give it a spin?

refrogerator commented:
cli sounds like an awesome idea! i'm currently trying to use the patcher on my steam deck and the gui is a pain because i have to reenter the port every time

RDIL commented:
@refrogerator it’s available in the current stable builds already.

LennardF1989 commented:
While the bundled version with the latest stable does include a CLI option, it does not have the Linux specific logic to make it work on the Steam Deck.

RDIL commented:
I think it would be best to switch to something lighter weight than winforms before merging would be best. Either that, or we’ll need to rethink our approach of committing the latest patcher binary for convenience.

LennardF1989 commented:
I'd say include the self-contained Linux CLI version (12MB) and the .NET Framework* based CLI and UI exe (200kb total). The UI has no use on Unix in any shape or form because you will have to run it through Wine in the same context as the game (which I couldn't even get working), and for Windows the .NET Framework one "just works".

Alternatively, we can just have a Unix installer script in a separate repository that grabs the CLI binary version whenever it's needed and keeps itself updated. Much like how installing a [insert package manager name] works on Unix.

* Only thing I want to adjust in that is making WinForms for .NET Framework DPI aware, so it's no longer blurry on high DPI monitors.

LennardF1989 commented:
It appears the patcher is unable to find the offsets on the latest version on Steam Deck. I'll see if I can get this fixed ASAP.

I also propose to pack both versions (original .NET and Unix compatible) with releases. That way the codebase is shared and potential updates to offsets affect all platforms accordingly.

Also contemplating adding some sort of UI support for Steam Deck which can supersede the bash scripts, but have to find the right tech to use for that.

AnthonyFuller commented:
3.210.0 introduced some compiler changes which inlined a lot and made most (if not all AOBs fail). You can see that some conflicts need to be merged, that's because the Windows patcher was updated to reflect these changes.

RDIL commented:
@AnthonyFuller I wouldn't say superseded... there should remain a no modding required alternative. Currently this "classic" version of the patcher serves that philosophy well enough. For that reason, I don't see us ever not shipping the patcher to make the experience as batteries included as possible.

As for the rollout of this new patcher, I'm not entirely sure how we want to proceed with it. My original thinking was:

- For a "transition period", ship the classic patcher with the zip, and the new patcher as a separate artifact in the releases
- After this, swap the two around
- Once we're confident the new patcher is at the maturity level that it needs to be, EOL the classic version

LennardF1989 commented:
Here's something I have been working on as a replacement for the current patcher UI.

![image](https://github.com/user-attachments/assets/844df038-aab2-487b-909f-1a7fcc71d9ba)

It uses SDL 3 (with software rendering) and ImGui and is cross-platform because of it. It uses the HitmanPatcher.Core I created as part of this PR, so effectively only replaces the WinForms UI. This means that Govert can keep things updated as it's still pretty much his code down to the core.

Best of all, when compiled and compressed into a single cross-platform executable, it's no bigger than 12-ish MB. A huge difference compared to the WinForms .NET Core version that is closer to 100MB.

SDL 3 also has support for tray icons, so that functionality will not be lost either.

LennardF1989 commented:
Can confirm it's working on the Steam Deck!

Some TODO's:
- BUG: MemoryPatcher can sometimes hang the Hitman process if it attempts to patch too soon after startup, need a way to detect a "not ready yet". There is a workaround however: start the Patcher after Hitman is actually working on the intro's.
- BUG: Looks like loading dynamic resources is no longer working, but could simply be my Peacock installation.
- Finalize UI: Scrollable log, maybe closable sections, add tray, add novikov status.

AnthonyFuller commented:
> BUG: Looks like loading dynamic resources is no longer working, but could simply be my Peacock installation.

This is known. IOI updated the client to not request dynamic resources. This is something the patcher needs to enable now (but the SDK can).

LennardF1989 commented:
@AnthonyFuller Is this already integrated into the SDK? If so, can you point me to the code that handles it?

AnthonyFuller commented:
Yes, it uses the [ZConfigCommand](https://github.com/OrfeasZ/ZHMModSDK/blob/master/ZHMModSDK/Include/Glacier/ZConfigCommand.h) class which is just what the `dynres_noforceoffline` and `configdomain` patches.

- [Optional dynamic resources (no force offline) example in the SDK](https://github.com/OrfeasZ/ZHMModSDK/blob/master/Mods/OnlineTools/src/OnlineTools.cpp#L128-L129)
- [Config domain example in the SDK](https://github.com/OrfeasZ/ZHMModSDK/blob/master/Mods/OnlineTools/src/OnlineTools.cpp#L131-L134)
- [Protocol example in the SDK](https://github.com/OrfeasZ/ZHMModSDK/blob/master/Mods/OnlineTools/src/OnlineTools.cpp#L137)

The SDK uses hooks, which I don't think you've added to this patcher. You could do that, or you could just use the AOB scanner to grab the required functions.

LennardF1989 commented:
Tried OnlineTools, checked all boxes, set to localhost. But even after restarting a few times I'm still getting [PEACOCK_...] on the main screen.

> IOI updated the client to not request dynamic resources
Maybe I'm misunderstanding, but this [DynRes patch](https://github.com/OrfeasZ/ZHMModSDK/blob/master/Mods/OnlineTools/src/OnlineTools.cpp#L128-L129) should fix that, right? Or is this something that is not implemented yet in OnlineTools?

AnthonyFuller commented:
I didn't say OnlineTools did it by default. I said that the SDK can do it, you need to type `config OnlineResources_Disable 0` into the console. Granted, it's not convenient, but that's because you can just install the Peacock Strings mod (and there's not much point updating the mod to include that, if anything saving custom values of config commands and applying them at launch is a better feature to spend time on).

> Maybe I'm misunderstanding, but this [DynRes patch](https://github.com/OrfeasZ/ZHMModSDK/blob/master/Mods/OnlineTools/src/OnlineTools.cpp#L128-L129) should fix that, right? Or is this something that is not implemented yet in OnlineTools?

That, as stated in the config command name, stops you from being forced offline when you fail to request dynamic resources. That is what the patcher currently does.

LennardF1989 commented:
Thanks, that's what I needed. Found the config offset, looking for a pattern now.

AnthonyFuller commented:
Awesome! I do realise I should've said that in my comment from yesterday but I made a bad assumption.

LennardF1989 commented:
No worries. Glad to report I got a patch working. Slightly annoying in the existing logic since it's an optional patch: if the offset can't be found, you are most likely running something that does not need it. Coincidentally: the actual offset in memory is +0x20 after the dynres_noforceoffline flag, so that could be a nice sanity check.

I'll create a separate PR for the existing patcher with this fix later.

VolkerSchlegel commented:
Great work, tested the Linux version on my setup, after adding in those three commits it worked perfectly:
https://github.com/thepeacockproject/Peacock/commit/dfd1bf3b2b2cd38500e12dacbe9715177ec374e0
https://github.com/thepeacockproject/Peacock/commit/5c9eb8eea34f0e522242c7655321c363c8fe639b
https://github.com/thepeacockproject/Peacock/commit/f2d5ce34b06e88b076d1d5cb7b0d1ad41f7f2230

thecatontheceiling commented:
I'd really love to see this merged. Would also be cool if this was made compatible with NativeAOT compilation somehow 👀
